### PR TITLE
Use only public API for EnterLeaveEventPlugin Tests

### PR DESCRIPTION
**Pull request by accordeiro**
This PR is related to issue #11299 – Express more tests via public API

I've managed to rewrite the `EnterLeavePlugin` tests using public API, but I'm not sure if I'm on the right path. What I did was basically analyze what was being done inside `EnterLeaveEventPlugin.js` and move the needed parts to the test (which luckily led to a much simpler code).

Does this make sense? Happy to make any necessary changes or follow a different path if need be.

Thanks for the opportunity of contributing! :)

### CODE REVIEW

**Comment by accordeiro** [comment]
Hi, @gaearon – thanks for the thorough feedback! :)

I've just pushed a new commit, trying to follow the path you've suggested. I've noticed two things:

1. When using `ReactTestUtils.SimulateNative`, the event's `relatedTarget` was set to `undefined`, so I had to set the relatedTarget manually, which made the last assertion a bit redundant. Am I missing something, or do you think this test makes sense anyway?

2. It seems that triggering the event via  `ReactTestUtils.SimulateNative` isn't actually exercising the `EnterLeavePlugin.js` codepath (I've put some logs on it to validate this assumption), while the second test on this file – which uses `ReactTestUtils.simulateNativeEventOnNode`–  actually exercises it. Should I rewrite the test to use this other event triggering function?

Thanks!

**Comment by gaearon** [comment]
I think I was wrong actually. `SimulateNative` is also not great because it bypasses the real browser events (which is what we want to test). Ideally we should just call `dispatchEvent` on nodes, like https://github.com/facebook/react/pull/11309 did.

**Comment by accordeiro** [comment]
OK, that makes sense – I've rewritten both tests to use `dispatchEvent` :)

File: packages/react-dom/src/events/__tests__/EnterLeaveEventPlugin-test.js [hunk 1]
```diff
@@ -9,56 +9,53 @@
...
-    var div = ReactDOM.findDOMNode(component);
 
-    var extracted = EnterLeaveEventPlugin.extractEvents(
-      'topMouseOver',
-      ReactDOMComponentTree.getInstanceFromNode(div),
-      {target: div},
-      div,
-    );
-    expect(extracted.length).toBe(2);
+    simulateMouseEvent(component, 'mouseover', iframe.contentWindow);
```
Comment (gaearon) [inline packages/react-dom/src/events/__tests__/EnterLeaveEventPlugin-test.js]: Shouldn't it be simulated on the `component`?
(which should be renamed to `node` btw)

I believe the original test did that.

File: packages/react-dom/src/events/__tests__/EnterLeaveEventPlugin-test.js [hunk 2]
```diff
@@ -9,56 +9,53 @@
...
-    var leave = extracted[0];
-    var enter = extracted[1];
-
-    expect(leave.target).toBe(iframe.contentWindow);
-    expect(leave.relatedTarget).toBe(div);
-    expect(enter.target).toBe(div);
-    expect(enter.relatedTarget).toBe(iframe.contentWindow);
+    expect(events.length).toBe(1);
+    expect(events[0].target).toBe(component);
+    expect(events[0].relatedTarget).toBe(iframe.contentWindow);
```
Comment (gaearon) [inline packages/react-dom/src/events/__tests__/EnterLeaveEventPlugin-test.js]: Which part of the code are this testing here? Are we confident that removing it causes test to fail?

**Comment by gaearon** [comment]
Specifically you can see the first test was added in https://github.com/facebook/react/commit/a2ecee5353d39be57ca49c6132eb0ca45d5be254.

This tells you *which* section of code it’s supposed to be testing. You can see that removing [these changes](https://github.com/facebook/react/commit/a2ecee5353d39be57ca49c6132eb0ca45d5be254#diff-7766218d52402c0ae54527f84a4d134aR84) doesn’t make the test fail, but it should.

You can also check that with the original test file, reverting those changes makes it hang. Which is not ideal (I'd prefer it fail 😛 ) but does demonstrate that we had coverage for this.

**Comment by gaearon** [comment]
Gentle ping, are you interested in following up on this?

**Comment by accordeiro** [comment]
Yes, of course – sorry for not following up on this before. I'll try to make the changes tomorrow at latest, is this ok? :)

**Comment by accordeiro** [comment]
Hi @gaearon – I've just pushed some changes. 

I've checked that deleting the code that you referenced [here](https://github.com/facebook/react/commit/a2ecee5353d39be57ca49c6132eb0ca45d5be254#diff-7766218d52402c0ae54527f84a4d134aR84) now makes the test fail. I've also tried setting the `win` var on the current implementation of `EnterLeavePlugin.js` to `null` makes the test fail.

I'm still having a hard time testing the `onMouseEnter` React event. I've tried dispatching both the `mouseout` and the `mouseenter` events (from opposite origins, naturally) and none of them have actually called the `onMouseEnter` function on the node. Is this something I should be worried about? 

Using `onMouseLeave`  + dispatching the `mouseleave` MouseEvent [like on the current diff] worked like a charm, though.

Please let me know if I'm on the right direction.

**Comment by gaearon** [comment]
>I'm still having a hard time testing the onMouseEnter React event. I've tried dispatching both the mouseout and the mouseenter events (from opposite origins, naturally) and none of them have actually called the onMouseEnter function on the node. Is this something I should be worried about?

You can see by the source that the only events React actually listens to in order to implement `onMouseEnter` (or `onMouseLeave`) is [`mouseover`](https://github.com/facebook/react/blob/221aa954fb2c2eb3b5d8fb52f9cc821db2ff1f7f/packages/react-dom/src/events/EnterLeaveEventPlugin.js#L41) and [`mouseout`](https://github.com/facebook/react/blob/221aa954fb2c2eb3b5d8fb52f9cc821db2ff1f7f/packages/react-dom/src/events/EnterLeaveEventPlugin.js#L46). Does this help?

**Comment by gaearon** [comment]
Thanks! This looks awesome.
I also added coverage for the `onMouseEnter` case which wasn't being tested by the new code.

**Pull request merged**